docs/core/diagnostics/diagnostics-in-containers.md

```diff
@@ -0,0 +1,69 @@
+---
+title: Collect diagnostics in containers
+description: In this article, you'll learn how .NET Core diagnostics tools can be used in Docker containers.
+ms.date: 09/01/2020
+---
+
+# Collect diagnostics in containers
+
+**This article applies to: ✔️** .NET Core 2.1 SDK and later versions
+
+The same diagnostics tools that are useful for diagnosing .NET Core issues in other scenarios also work in Docker containers. However, some of the tools require special steps to work in a container. This article covers how tools for gathering performance traces and collecting dumps can be used in Docker containers.
+
+## Using .NET Core global CLI tools in a container
+
+The .NET Core global CLI diagnostic tools ([`dotnet-counters`](dotnet-counters.md), [`dotnet-dump`](dotnet-dump.md), [`dotnet-gcdump`](dotnet-gcdump.md), and [`dotnet-trace`](dotnet-trace.md)) are designed to work in a wide variety of environments and should all work directly in Docker containers. Because of this, these tools are the preferred method of collecting diagnostic information for .NET Core scenarios targeting .NET Core 3.0 or above (or 3.1 or above in the case of `dotnet-gcdump`) in containers.
+
+The only complicating factor of using these tools in a container is that they are installed with the .NET Core SDK and many Docker containers run without the .NET Core SDK present. One easy solution to this problem is to install the tools in the initial Docker image. The tools don't need the .NET Core SDK to run, only to be installed. Therefore, it's possible to create a Dockerfile that installs the tools in a build stage (where the .NET Core SDK is present) and then copies the binaries into the final image. The only downside to this approach is increased Docker image size (each tool is roughly 100 KB, give or take, depending on the specific tool and OS).
+
+```dockerfile
+# In build stage
+# Install desired .NET CLI diagnostics tools
+RUN dotnet tool install --tool-path /tools dotnet-trace
+RUN dotnet tool install --tool-path /tools dotnet-counters
+RUN dotnet tool install --tool-path /tools dotnet-dump
+
+...
+
+# In final stage
+# Copy diagnostics tools
+WORKDIR /tools
+COPY --from=build /tools .
+```
+
+Alternatively, the .NET Core SDK can be installed in a container when needed in order to install the CLI tools. Be aware that installing the .NET Core SDK will have the side-effect of reinstalling the .NET Core runtime. So be sure to install the version of the SDK that matches the runtime present in the container.
+
+### Using .NET Core global CLI tools in a sidecar container
+
+If you would like to use .NET Core global CLI diagnostic tools to diagnose processes in a different container, bear the following additional requirements in mind:
+
+1. The containers must, of course, [share a process namespace](https://docs.docker.com/engine/reference/run/#pid-settings---pid) (so that tools in the sidecar container can access processes in the target container).
```

Comment:
I don't now if it's worth calling out, but you can definitely use the diagnostics tools without sharing process namespace. I'm okay leaving this as is for now, and updating it in a future iteration once the tool can successfully consume the new process info commands.

@josalem Thoughts?

Reply:
The socket discovery mechanics fall apart if you aren't sharing a PID-namespace and there are multiple container's apps you want to connect to, e.g., I have containers A, B, and C (my diag container).  If you don't share a process namespace, but you share `/tmp`, you'll end up multiple sockets that only differentiate by their epoch timestamp.  True we could use the process info command, but that only exists in 5.0 and there is a better mechanism than sharing `/tmp` in 5.0.

If you only have 1 container you want to connect to, then not sharing a PID-space is okay.  I think this nuance should be called out in the text if we remove the suggestion to share a PID-space.

These docs should include instructions that are limited to "best experience" for 3.1 I think.  We can add 5.0 specific instructions once that ships and the tools have finished being updated to use the new features.

Reply:
In the usual use case, won't customers need to share PID-space so that the `-p` parameters will work? If there's a reasonably straight-forward way to trace/watch/dump from another container without sharing process namespace, I can include that, but it might be better to wait and update in the .NET 5 timeframe with the updated experience.